// mineflayer/lib/plugins/generic_place.js
const assert = require('assert')
module.exports = inject

function inject (bot) {
  const Item = require('prismarine-item')(bot.registry)
  const mcData = require('minecraft-data')('1.15.2')
  const specialPlace = {
      'dirt': ['grass', 8],
      'grass': ['dirt', 9], 
      'granite': ['white_concrete_powder', 429],
      'white_concrete_powder': ['white_concrete_powder_stairs', 884],
      'white_concrete_powder_stairs': ['white_concrete_powder', 429],
      'orange_concrete_powder_stairs': ['orange_concrete_powder', 430],
      'orange_concrete_powder': ['orange_concrete_powder_stairs', 681],
      'magenta_concrete_powder_stairs': ['magenta_concrete_powder', 431],
      'magenta_concrete_powder': ['magenta_concrete_powder_stairs', 682],
      'light_blue_concrete_powder_stairs': ['light_blue_concrete_powder', 432],
      'light_blue_concrete_powder': ['light_blue_concrete_powder_stairs', 683],
      'yellow_concrete_powder_stairs': ['yellow_concrete_powder', 433],
      'yellow_concrete_powder': ['yellow_concrete_powder_stairs', 684],
      'lime_concrete_powder_stairs': ['lime_concrete_powder', 434],
      'lime_concrete_powder': ['lime_concrete_powder_stairs', 685],
      'pink_concrete_powder_stairs': ['pink_concrete_powder', 435],
      'pink_concrete_powder': ['pink_concrete_powder_stairs', 686],
      'gray_concrete_stairs': ['gray_concrete', 420],
      'gray_concrete': ['gray_concrete_stairs', 687],
      'light_gray_concrete_stairs': ['light_gray_concrete', 421],
      'light_gray_concrete': ['light_gray_concrete_stairs', 688],
      'cyan_concrete_stairs': ['cyan_concrete', 422],
      'cyan_concrete': ['cyan_concrete_stairs', 689],
      'purple_concrete_stairs': ['purple_concrete', 423],
      'purple_concrete': ['purple_concrete_stairs', 690],
      'blue_concrete_stairs': ['blue_concrete', 424],
      'blue_concrete': ['blue_concrete_stairs', 691],
      'brown_concrete_stairs': ['brown_concrete', 425],
      'brown_concrete': ['brown_concrete_stairs', 692],
      'green_concrete_stairs': ['green_concrete', 426],
      'green_concrete': ['green_concrete_stairs', 693],
      'red_concrete_stairs': ['red_concrete', 427],
      'red_concrete': ['red_concrete_stairs', 694],
      'black_concrete_stairs': ['black_concrete', 428],
      'black_concrete': ['black_concrete_stairs', 695]
  }
  /**
   *
   * @param {import('prismarine-block').Block} referenceBlock
   * @param {import('vec3').Vec3} faceVector
   * @param {{half?: 'top'|'bottom', delta?: import('vec3').Vec3, forceLook?: boolean | 'ignore', offhand?: boolean, swingArm?: 'right' | 'left', showHand?: boolean}} options
   */
  async function _genericPlace (referenceBlock, faceVector, options) {
    let handToPlaceWith = 0
    let item = bot.heldItem
    if (options.offhand) {
      if (!bot.inventory.slots[45]) {
        throw new Error('must be holding an item in the off-hand to place')
      }
      handToPlaceWith = 1
    } else if (!item) {
      throw new Error('must be holding an item to place')
    }

    // Look at the center of the face
    let dx = 0.5 + faceVector.x * 0.5
    let dy = 0.5 + faceVector.y * 0.5
    let dz = 0.5 + faceVector.z * 0.5
    if (dy === 0.5) {
      if (options.half === 'top') dy += 0.25
      else if (options.half === 'bottom') dy -= 0.25
    }
    if (options.delta) {
      dx = options.delta.x
      dy = options.delta.y
      dz = options.delta.z
    }
    if (options.forceLook !== 'ignore') {
      await bot.lookAt(referenceBlock.position.offset(dx, dy, dz), options.forceLook)
    }
    // TODO: tell the server that we are sneaking while doing this
    const pos = referenceBlock.position

    if (options.swingArm) {
      bot.swingArm(options.swingArm, options.showHand)
    }

    console.log("Item is " + item.name)
    if (options.ctrlKey && specialPlace.hasOwnProperty(item.name)){
      console.log("Control place used")
      // use hand to pass on the fact it's a strange block
      handToPlaceWith = 2
    }

    if (bot.supportFeature('blockPlaceHasHeldItem')) {
      console.log("Place type 1")
      const packet = {
        location: pos,
        direction: vectorToDirection(faceVector),
        heldItem: Item.toNotch(item),
        hand: handToPlaceWith,
        cursorX: Math.floor(dx * 16),
        cursorY: Math.floor(dy * 16),
        cursorZ: Math.floor(dz * 16),
      }
      bot._client.write('block_place', packet)
    } else if (bot.supportFeature('blockPlaceHasHandAndIntCursor')) {
      console.log("Place type 2")
      bot._client.write('block_place', {
        location: pos,
        direction: vectorToDirection(faceVector),
        hand: handToPlaceWith,
        cursorX: Math.floor(dx * 16),
        cursorY: Math.floor(dy * 16),
        cursorZ: Math.floor(dz * 16),
      })
    } else if (bot.supportFeature('blockPlaceHasHandAndFloatCursor')) {
      console.log("Place type 3")
      bot._client.write('block_place', {
        location: pos,
        direction: vectorToDirection(faceVector),
        hand: handToPlaceWith,
        cursorX: dx,
        cursorY: dy,
        cursorZ: dz,
      })
    } else if (bot.supportFeature('blockPlaceHasInsideBlock')) {
      console.log("Place type 4")
      bot._client.write('block_place', {
        location: pos,
        direction: vectorToDirection(faceVector),
        hand: handToPlaceWith,
        cursorX: dx,
        cursorY: dy,
        cursorZ: dz,
        insideBlock: false,
      })
    }

    return pos
  }
  bot._genericPlace = _genericPlace
}

function vectorToDirection (v) {
  if (v.y < 0) {
    return 0
  } else if (v.y > 0) {
    return 1
  } else if (v.z < 0) {
    return 2
  } else if (v.z > 0) {
    return 3
  } else if (v.x < 0) {
    return 4
  } else if (v.x > 0) {
    return 5
  }
  assert.ok(false, `invalid direction vector ${v}`)
}
