// mineflayer/lib/plugins/generic_place.js
const assert = require('assert')
module.exports = inject

function inject (bot) {
  const Item = require('prismarine-item')(bot.registry)
  const mcData = require('minecraft-data')('1.15.2')
    const specialPlace = [
      'dirt', 'grass_block', 
      'white_concrete_stairs', 'orange_concrete_stairs', 'magenta_concrete_stairs',
      'light_blue_concrete_stairs', 'yellow_concrete_stairs', 'lime_concrete_stairs',
      'pink_concrete_stairs', 'gray_concrete_stairs', 'light_gray_concrete_stairs',
      'cyan_concrete_stairs', 'purple_concrete_stairs', 'blue_concrete_stairs',
      'brown_concrete_stairs', 'green_concrete_stairs', 'red_concrete_stairs',
      'black_concrete_stairs',
      'white_concrete', 'orange_concrete', 'magenta_concrete',
      'light_blue_concrete', 'yellow_concrete_powder', 'lime_concrete',
      'pink_concrete_powder', 'gray_concrete', 'light_gray_concrete',
      'cyan_concrete', 'purple_concrete', 'blue_concrete',
      'brown_concrete', 'green_concrete', 'red_concrete',
      'black_concrete',
    ]
  /**
   *
   * @param {import('prismarine-block').Block} referenceBlock
   * @param {import('vec3').Vec3} faceVector
   * @param {{half?: 'top'|'bottom', delta?: import('vec3').Vec3, forceLook?: boolean | 'ignore', offhand?: boolean, swingArm?: 'right' | 'left', showHand?: boolean}} options
   */
  async function _genericPlace (referenceBlock, faceVector, options) {
    let handToPlaceWith = 0
    let item = bot.heldItem
    if (options.offhand) {
      if (!bot.inventory.slots[45]) {
        throw new Error('must be holding an item in the off-hand to place')
      }
      handToPlaceWith = 1
    } else if (!item) {
      throw new Error('must be holding an item to place')
    }

    // Look at the center of the face
    let dx = 0.5 + faceVector.x * 0.5
    let dy = 0.5 + faceVector.y * 0.5
    let dz = 0.5 + faceVector.z * 0.5
    if (dy === 0.5) {
      if (options.half === 'top') dy += 0.25
      else if (options.half === 'bottom') dy -= 0.25
    }
    if (options.delta) {
      dx = options.delta.x
      dy = options.delta.y
      dz = options.delta.z
    }
    if (options.forceLook !== 'ignore') {
      await bot.lookAt(referenceBlock.position.offset(dx, dy, dz), options.forceLook)
    }
    // TODO: tell the server that we are sneaking while doing this
    const pos = referenceBlock.position

    if (options.swingArm) {
      bot.swingArm(options.swingArm, options.showHand)
    }

    console.log("Item is " + item.name)
    if (options.ctrlKey && specialPlace.includes(item.name)){
      console.log("Control place used")
      // use hand to pass on the fact it's a strange block
      handToPlaceWith = 2
    }

    if (bot.supportFeature('blockPlaceHasHeldItem')) {
      console.log("Place type 1")
      const packet = {
        location: pos,
        direction: vectorToDirection(faceVector),
        heldItem: Item.toNotch(item),
        hand: handToPlaceWith,
        cursorX: Math.floor(dx * 16),
        cursorY: Math.floor(dy * 16),
        cursorZ: Math.floor(dz * 16),
      }
      bot._client.write('block_place', packet)
    } else if (bot.supportFeature('blockPlaceHasHandAndIntCursor')) {
      console.log("Place type 2")
      bot._client.write('block_place', {
        location: pos,
        direction: vectorToDirection(faceVector),
        hand: handToPlaceWith,
        cursorX: Math.floor(dx * 16),
        cursorY: Math.floor(dy * 16),
        cursorZ: Math.floor(dz * 16),
      })
    } else if (bot.supportFeature('blockPlaceHasHandAndFloatCursor')) {
      console.log("Place type 3")
      bot._client.write('block_place', {
        location: pos,
        direction: vectorToDirection(faceVector),
        hand: handToPlaceWith,
        cursorX: dx,
        cursorY: dy,
        cursorZ: dz,
      })
    } else if (bot.supportFeature('blockPlaceHasInsideBlock')) {
      console.log("Place type 4")
      bot._client.write('block_place', {
        location: pos,
        direction: vectorToDirection(faceVector),
        hand: handToPlaceWith,
        cursorX: dx,
        cursorY: dy,
        cursorZ: dz,
        insideBlock: false,
      })
    }

    return pos
  }
  bot._genericPlace = _genericPlace
}

function vectorToDirection (v) {
  if (v.y < 0) {
    return 0
  } else if (v.y > 0) {
    return 1
  } else if (v.z < 0) {
    return 2
  } else if (v.z > 0) {
    return 3
  } else if (v.x < 0) {
    return 4
  } else if (v.x > 0) {
    return 5
  }
  assert.ok(false, `invalid direction vector ${v}`)
}
